src/data/api-service.ts

```diff
@@ -34,14 +35,61 @@ Please check the url: ${config.baseURL}${config.url}`
 }
 
 /**
- *
- * @param baseUrl - the optional base url to override the `env.apiUrl`
+ * the list of options that can be passed when instantiating a new API service.
  */
-export const createApiService = (baseUrl = process.env.apiUrl) => {
-  const client = axios.create({
-    baseURL: baseUrl,
+export interface ApiServiceConfig {
+  /** to use a different base URL than configured for the environment */
+  baseUrl?: string
+  /** to make authenticated requests to the API */
+  accessToken?: string
+  /** whether to use the `'v1/'` prefix after the base URL */
+  isVersioned?: boolean
+}
+
+/**
+ * the schema of the API service
+ */
+export interface ApiService {
+  query<T = unknown>(
+    resource: string,
+    params: Record<string, string>
+  ): Promise<AxiosResponse<T>>
+  get<T = unknown>(resource: string, slug: string): Promise<AxiosResponse<T>>
+  post<T = unknown>(
+    resource: string,
+    data: Parameters<AxiosInstance['post']>[1],
+    headers?: AxiosRequestConfig['headers']
+  ): Promise<AxiosResponse<T>>
+  update<T = unknown>(
+    resource: string,
+    slug: string,
+    data: Parameters<AxiosInstance['put']>[1],
+    headers: AxiosRequestConfig['headers']
+  ): Promise<AxiosResponse<T>>
+  put<T = unknown>(
+    resource: string,
+    params: AxiosRequestConfig
+  ): Promise<AxiosResponse<T>>
+  delete<T = unknown>(
+    resource: string,
+    slug: string,
+    headers: AxiosRequestConfig['headers']
+  ): Promise<AxiosResponse<T>>
+}
+
+export const createApiService = ({
+  baseUrl = process.env.apiUrl,
+  accessToken = undefined,
+  isVersioned = true,
+}: ApiServiceConfig = {}): ApiService => {
+  const axiosParams: AxiosRequestConfig = {
+    baseURL: isVersioned ? `${baseUrl}v1/` : baseUrl,
```

Comment:
I think `isVersioned` here assumes that the api service being created will only be used for a particular set of resources or requests rather than all requests. Because we create multiple axios clients for different api service usages, we pass `isVersioned` when we want the URLs to automatically include the v1 string.

However, I also disagree with this change. @dhruvkb can you explain the benefit of this implementation over explicitly including the version string at the request site? If we introduce new versions of existing endpoints or have api services used for mixed version requests in the same cycle then this implementation will either have to be reverted back to the explicit version or be significantly increase the complexity of this code.

I prefer explicit request URLs personally, they are easier to search for in the app and make zero assumptions about the future or current versions of API endpoints that will or are being used.

Reply:
Well, I think undoing this refactor is actually more trouble than it is worth and will complicate this PR. The existing "versioned API" approach we took was effectively the same and had the same bad assumptions (that `v1` is the only version and that all requests performed by a given API service instance will all reference the same version of endpoint).

I personally find the current approach with the `getResourceSlug` function over-complicated, sneaky, and bakes assumptions into the base `api-service.ts` module that are not necessarily true.

Either the API service instances are only for making requests to the media endpoints (and we should use a plain axios instance for anything else) or we have to reconsider how the API service is used in the first place.

I think it's potentially an anti-pattern, the way it's currently implemented, and we might want to consider stuffing it into Nuxt context, either by hand-spinning our own module or using the existing nuxt/axios module.

This would have some significant benefits:

1. We could insert the plugin after the API token request plugin so that it can automatically include the access token on the server. This allows downstream requesters to completely ignore the api token detail. Currently all api service instantiators have to consider the api token. This duplicates the work and complicates the downstream code.
2. We can eliminate the complexity in the `api-service` module completely. The only issue that remains is the mapping of the media type to the endpoint names, but this can easily be put into a function or a static mapping for reference by requesters to construct API urls.

I'll open an issue for this. For now I think this should remain as is with @dhruvkb's refactors because it keeps the existing assumptions, but makes them ever so slightly more flexible to accommodate non-versioned API requests.

Reply:
https://github.com/WordPress/openverse-frontend/issues/1591